src/controllers/GovernanceTimelock.sol

```diff
@@ -0,0 +1,158 @@
+                                MODIFIERS
+    ////////////////////////////////////////////////////////////////*/
+
+    /// @notice A modifier used for restricting the caller of some functions to be this contract itself.
+    modifier onlySelf() {
+        if (msg.sender != address(this)) {
+            revert Error.INVALID_SELF_CALLER();
+        }
+        _;
+    }
+
+    /// @notice A modifier used for restricting caller to admin contract
+    modifier onlyAdmin() {
+        if (msg.sender != admin) {
+            revert Error.CALLER_NOT_ADMIN();
+        }
+        _;
+    }
+
+    /*/////////////////////////////////////////////////////////////////
+                                CONSTRUCTOR
+    ////////////////////////////////////////////////////////////////*/
+
+    /// @param _admin is the address of admin contract that schedule txs
+    constructor(address _admin) {
+        if (_admin == address(0)) {
+            revert Error.ZERO_ADDRESS_INPUT();
+        }
+
+        admin = _admin;
+        emit AdminUpdated(address(0), _admin);
+    }
+
+    /*/////////////////////////////////////////////////////////////////
+                            EXTERNAL FUNCTIONS
+    ////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IGovernanceTimelock
+    function scheduleTransaction(address _target, uint256 _value, bytes memory _data) external override onlyAdmin {
+        if (_target == address(0)) {
+            revert Error.INVALID_TARGET();
+        }
+
+        /// increment tx counter
+        ++txCounter;
+        uint256 eta = block.timestamp + delay;
+
+        scheduledTransaction[txCounter] = keccak256(abi.encodePacked(_target, _value, _data, eta));
+        emit TransactionScheduled(txCounter, _target, _value, _data, eta);
+    }
+
+    /// @inheritdoc IGovernanceTimelock
+    function executeTransaction(uint256 _txId, address _target, uint256 _value, bytes memory _data, uint256 _eta)
+        external
+        payable
+        override
+    {
+        /// @dev validates the txId
+        if (_txId == 0 || _txId > txCounter) {
+            revert Error.INVALID_TX_ID();
+        }
+
+        /// @dev checks if tx is already executed;
+        if (isExecuted[_txId]) {
+            revert Error.TX_ALREADY_EXECUTED();
+        }
+
+        /// @dev check the input params against hash
+        if (scheduledTransaction[_txId] != keccak256(abi.encodePacked(_target, _value, _data, _eta))) {
+            revert Error.INVALID_TX_INPUT();
+        }
+
+        /// @dev checks timelock
+        if (_eta > block.timestamp) {
+            revert Error.TX_TIMELOCKED();
+        }
+
+        /// @dev checks if tx within execution period
+        if (block.timestamp > _eta + GRACE_PERIOD) {
+            revert Error.TX_EXPIRED();
+        }
+
+        /// @dev checks native funding
+        if (msg.value != _value) {
+            revert Error.TX_UNDERPAID();
```

Comment:
nit: technically `msg.value` could be higher than `value` when this condition is triggered, but that would not really imply "underpaid"